File: rust/crates/agent-runtime/src/workspaces/git.rs
use std::ffi::OsStr;
use std::path::Path;

use anyhow::{bail, Context, Result};
use tokio::process::Command;

use super::config::{required_git_field, WorkspaceBaseConfig};

pub(super) async fn refresh_git_workspace_base(
    base: &Path,
    config: &WorkspaceBaseConfig,
) -> Result<()> {
    let remote_url = required_git_field(config.remote_url.as_deref(), "remote_url")?;
    let branch = required_git_field(config.remote_branch.as_deref(), "remote_branch")?;
    let branch_refspec = format!("+refs/heads/{branch}:refs/remotes/origin/{branch}");

    if !base.join(".git").is_dir() {
        run_git(base, ["init"]).await?;
    }

    if git_remote_exists(base, "origin").await? {
        run_git(base, ["remote", "set-url", "origin", remote_url]).await?;
    } else {
        run_git(base, ["remote", "add", "origin", remote_url]).await?;
    }
    run_git(base, ["fetch", "--prune", "origin", &branch_refspec]).await?;
    let origin_ref = format!("refs/remotes/origin/{branch}");
    let base_sha = git_output(base, ["rev-parse", &origin_ref]).await?;
    run_git(base, ["checkout", "--detach", &base_sha]).await?;
    run_git(base, ["reset", "--hard", &base_sha]).await?;
    run_git(base, ["clean", "-ffdx"]).await?;
    Ok(())
}

/// Fetch a per-session branch override into an already-instantiated git workspace
/// and return its commit sha.
///
/// The session workspace inherits `origin` from the managed project base, so this
/// only needs to fetch the override branch (the project base itself stays on the
/// project's configured branch and is shared across sessions). Errors when the
/// branch name is invalid or absent on the remote; these are client-input errors.
pub(super) async fn fetch_session_branch_head(workspace: &Path, branch: &str) -> Result<String> {
    let branch = branch.trim();
    if branch.is_empty() {
        bail!("session branch override is required");
    }
    let branch_check = git_command()
        .args(["check-ref-format", "--branch", branch])
        .output()
        .await
        .context("validate session branch override name")?;
    if !branch_check.status.success() {
        bail!("session branch override is not a valid git branch name: {branch}");
    }
    let branch_refspec = format!("+refs/heads/{branch}:refs/remotes/origin/{branch}");
    let fetch = git_command()
        .args(["fetch", "--prune", "origin", &branch_refspec])
        .current_dir(workspace)
        .output()
        .await
        .with_context(|| format!("fetch session branch {branch} in {}", workspace.display()))?;
    if !fetch.status.success() {
        bail!(
            "session branch override not found on remote: {branch}: {}",
            String::from_utf8_lossy(&fetch.stderr).trim()
        );
    }
    let origin_ref = format!("refs/remotes/origin/{branch}");
    git_output(workspace, ["rev-parse", &origin_ref]).await
}

pub async fn validate_remote_branch(remote_url: &str, remote_branch: &str) -> Result<()> {
    let remote_url = remote_url.trim();
    let remote_branch = remote_branch.trim();
    if remote_url.is_empty() {
        bail!("workspace remote_url is required");
    }
    if remote_branch.is_empty() {
        bail!("workspace remote_branch is required");
    }
    let branch_check = git_command()
        .arg("check-ref-format")
        .arg("--branch")
        .arg(remote_branch)
        .output()
        .await
        .context("validate remote branch name")?;
    if !branch_check.status.success() {
        bail!("workspace remote_branch is not a valid git branch name: {remote_branch}");
    }
    let output = git_command()
        .arg("ls-remote")
        .arg("--heads")
        .arg(remote_url)
        .arg(remote_branch)
        .output()
        .await
        .with_context(|| format!("check remote branch {remote_url} {remote_branch}"))?;
    if !output.status.success() {
        bail!(
            "git ls-remote failed for {remote_url} {remote_branch}: {}",
            String::from_utf8_lossy(&output.stderr).trim()
        );
    }
    if String::from_utf8_lossy(&output.stdout).trim().is_empty() {
        bail!("remote branch not found: {remote_url} {remote_branch}");
    }
    Ok(())
}

/// Run `git` in `cwd` (optionally against an alternate index) and return trimmed
/// stdout. The typed wrappers below choose the arg shape and whether stdout is
/// used; this is the single place that spawns git and maps failure to an error.
async fn git(cwd: &Path, args: impl IntoIterator<Item = impl AsRef<OsStr>>) -> Result<String> {
    let output = git_command()
        .args(args)
        .current_dir(cwd)
        .output()
        .await
        .with_context(|| format!("run git in {}", cwd.display()))?;
    if !output.status.success() {
        bail!(
            "git failed in {}: {}",
            cwd.display(),
            String::from_utf8_lossy(&output.stderr).trim()
        );
    }
    Ok(String::from_utf8_lossy(&output.stdout).trim().to_string())
}

pub(super) async fn run_git<const N: usize>(cwd: &Path, args: [&str; N]) -> Result<()> {
    git(cwd, args).await.map(|_| ())
}

pub(super) async fn git_output<const N: usize>(cwd: &Path, args: [&str; N]) -> Result<String> {
    git(cwd, args).await
}

async fn git_remote_exists(cwd: &Path, name: &str) -> Result<bool> {
    let output = git_command()
        .args(["remote", "get-url", name])
        .current_dir(cwd)
        .output()
        .await
        .with_context(|| format!("check git remote {name} in {}", cwd.display()))?;
    Ok(output.status.success())
}

fn git_command() -> Command {
    let mut command = Command::new("git");
    command
        .env("GIT_TERMINAL_PROMPT", "0")
        .env("GIT_AUTHOR_NAME", "pi-relay")
        .env("GIT_AUTHOR_EMAIL", "pi-relay@example.invalid")
        .env("GIT_COMMITTER_NAME", "pi-relay")
        .env("GIT_COMMITTER_EMAIL", "pi-relay@example.invalid");
    command
}

File: packages/workspace-lib/src/git.ts
// Git base refresh + session branch ops (port of workspaces/git.rs).
import { existsSync } from "node:fs";
import { WorkspaceError } from "./errors.ts";
import { run } from "./exec.ts";
import type { WorkspaceBaseConfig } from "./config.ts";

function gitEnv(): NodeJS.ProcessEnv {
	return {
		...process.env,
		GIT_TERMINAL_PROMPT: "0",
		GIT_AUTHOR_NAME: "pi-relay",
		GIT_AUTHOR_EMAIL: "pi-relay@example.invalid",
		GIT_COMMITTER_NAME: "pi-relay",
		GIT_COMMITTER_EMAIL: "pi-relay@example.invalid",
	};
}

async function git(cwd: string, args: string[]): Promise<string> {
	const res = await run("git", args, { cwd, env: gitEnv() });
	return res.stdout.trim();
}

async function gitRemoteExists(cwd: string, name: string): Promise<boolean> {
	try {
		await git(cwd, ["remote", "get-url", name]);
		return true;
	} catch {
		return false;
	}
}

/** refresh_git_workspace_base: init if needed, fetch the configured branch,
 * hard-reset the base to origin/<branch>, clean -ffdx. Base stays detached. */
export async function refreshGitWorkspaceBase(base: string, config: WorkspaceBaseConfig): Promise<void> {
	const remoteUrl = config.remote_url ?? "";
	const remoteBranch = config.remote_branch ?? "";
	const branchRefspec = `+refs/heads/${remoteBranch}:refs/remotes/origin/${remoteBranch}`;
	if (!existsSync(`${base}/.git`)) {
		await git(base, ["init"]);
	}
	if (await gitRemoteExists(base, "origin")) {
		await git(base, ["remote", "set-url", "origin", remoteUrl]);
	} else {
		await git(base, ["remote", "add", "origin", remoteUrl]);
	}
	await git(base, ["fetch", "--prune", "origin", branchRefspec]);
	const originRef = `refs/remotes/origin/${remoteBranch}`;
	const baseSha = await git(base, ["rev-parse", originRef]);
	await git(base, ["checkout", "--detach", baseSha]);
	await git(base, ["reset", "--hard", baseSha]);
	await git(base, ["clean", "-ffdx"]);
}

/** fetch_session_branch_head: fetch a per-session branch override into an
 * already-materialized git workspace; returns the commit sha. */
export async function fetchSessionBranchHead(workspace: string, branch: string): Promise<string> {
	const trimmed = branch.trim();
	if (trimmed === "") throw new WorkspaceError("bad_workspace", "session branch override is required");
	try {
		await run("git", ["check-ref-format", "--branch", trimmed], { env: gitEnv() });
	} catch {
		throw new WorkspaceError("bad_workspace", `session branch override is not a valid git branch name: ${trimmed}`);
	}
	const branchRefspec = `+refs/heads/${trimmed}:refs/remotes/origin/${trimmed}`;
	try {
		await run("git", ["fetch", "--prune", "origin", branchRefspec], { cwd: workspace, env: gitEnv() });
	} catch {
		throw new WorkspaceError("bad_workspace", `session branch override not found on remote: ${trimmed}`);
	}
	return git(workspace, ["rev-parse", `refs/remotes/origin/${trimmed}`]);
}

export async function validateRemoteBranch(remoteUrl: string, remoteBranch: string): Promise<void> {
	const url = remoteUrl.trim();
	const branch = remoteBranch.trim();
	if (url === "") throw new WorkspaceError("bad_workspace", "workspace remote_url is required");
	if (branch === "") throw new WorkspaceError("bad_workspace", "workspace remote_branch is required");
	try {
		await run("git", ["check-ref-format", "--branch", branch], { env: gitEnv() });
	} catch {
		throw new WorkspaceError("bad_workspace", `workspace remote_branch is not a valid git branch name: ${branch}`);
	}
	const res = await run("git", ["ls-remote", "--heads", url, branch], { env: gitEnv() });
	if (res.stdout.trim() === "") {
		throw new WorkspaceError("bad_workspace", `remote branch not found: ${url} ${branch}`);
	}
}

export async function revParse(cwd: string, ref: string): Promise<string> {
	return git(cwd, ["rev-parse", ref]);
}

/** checkout_session_branch: create (or reset) the per-session branch at sha. */
export async function checkoutSessionBranch(workspace: string, branch: string, sha: string): Promise<void> {
	await git(workspace, ["checkout", "-B", branch, sha]);
}

export async function mergeBase(workspace: string, a: string, b: string): Promise<string> {
	return git(workspace, ["merge-base", a, b]);
}

export async function currentBranch(workspace: string): Promise<string | null> {
	const name = await git(workspace, ["branch", "--show-current"]);
	return name === "" ? null : name;
}
